blockchain/validate_test.go

```diff
@@ -313,66 +315,69 @@ func TestSigOpsLimitsWithMultiMBBlocks(t *testing.T) {
+	// Add 3 * 400k NOOPs to get >1MB block without breaking Tx size limits
+	manyNoOps := repeatScript(400000, txscript.OP_NOP)
+	for _, tx := range baseBlock.Transactions {
+		tx.TxOut[0].PkScript = append(tx.TxOut[0].PkScript, manyNoOps...)
+	}
 
 	// 1. Pass at per-Tx sigOps limit
 	// coinbase:     20k sigOps (ok)     <=1MB (ok)
 	// tx1:            1 sigOps (ok)     <=1MB (ok)
 	// tx2:            1 sigOps (ok)     <=1MB (ok)
 	// block:    20k + 2 sigOps (ok)     > 1MB (ok)
-	baseBlock.Transactions[0].TxOut[0].PkScript = bigEnoughScript
 	atTxLimit, err := newTestBlock(baseBlock, tip, 20000, 1, 1)
 	if err != nil {
 		t.Fatalf("Unexpected error creating sigOps test block: %v", err)
 	}
-	err = attachBigBlock(atTxLimit)
+	err = validateBigBlock(atTxLimit)
 	if err != nil {
-		t.Fatalf("Expected to validate but got: %v", err)
+		if rule, ok := err.(RuleError); ok {
+			if rule.ErrorCode == ErrTooManySigOps {
+				t.Fatalf("Expected to validate but got: %v", err)
+			}
+		}
+
 	}
 	tip = atTxLimit
 
 	// 1. Fail at per-Tx sigOps limit +1
-	// coinbase: 20k + 1 sigOps (fail)   <=1MB (ok)
+	// coinbase: 40k + 1 sigOps (fail)   <=1MB (ok)
 	// tx1:            1 sigOps (ok)     <=1MB (ok)
 	// tx2:            1 sigOps (ok)     <=1MB (ok)
-	// block:    20k + 3 sigOps (ok)     > 1MB (ok)
-	fmt.Println("1. Fail")
-	overTxLimit, err := newTestBlock(baseBlock, tip, 20001, 1, 1)
+	// block:    40k + 3 sigOps (ok)     > 1MB (ok)
+	overTxLimit, err := newTestBlock(baseBlock, tip, 40001, 1, 1)
 	if err != nil {
 		t.Fatalf("Unexpected error creating sigOps test block: %v", err)
 	}
-	err = attachBigBlock(overTxLimit)
-	isExpectedErr := false
+	err = validateBigBlock(overTxLimit)
 	if err != nil {
 		if rule, ok := err.(RuleError); ok {
-			if rule.ErrorCode == ErrTxTooManySigOps {
-				isExpectedErr = true
+			if rule.ErrorCode != ErrTooManySigOps {
+				t.Fatalf("Expected to fail with TooManySigOps but got: %v", err)
 			}
 		}
 	}
-	if !isExpectedErr {
-		t.Fatalf("Expected to fail with TxTooManySigOps but got: %v", err)
-	}
 
 	// 2. Pass at per-block sigOps limit
 	// coinbase:     20k sigOps (ok)     <=1MB (ok)
 	// tx1:        19999 sigOps (ok)     <=1MB (ok)
 	// tx2:            1 sigOp  (ok)     <=1MB (ok)
 	// block:        40k sigOps (ok)     > 1MB (ok)
-	bigEnoughScript = append(baseScriptForSize, repeatScript(959800, txscript.OP_NOP)...)
-	baseBlock.Transactions[0].TxOut[0].PkScript = bigEnoughScript
 	atBlockLimit, err := newTestBlock(baseBlock, tip, 20000, 19999, 1)
 	if err != nil {
 		t.Fatalf("Unexpected error creating sigOps test block: %v", err)
 	}
-	err = attachBigBlock(atBlockLimit)
+	err = validateBigBlock(atBlockLimit)
 	if err != nil {
-		t.Fatalf("Expected to validate but got: %v", err)
+		if rule, ok := err.(RuleError); ok {
+			if rule.ErrorCode == ErrTooManySigOps {
+				t.Fatalf("Expected to validate but got: %v", err)
+			}
+		}
+
 	}
-	tip = atBlockLimit
+	tip = atTxLimit
```

Comment:
This was a bug in the old code. The tip needs to point to `atBlockLimit` here since it should be attached in the previous step.

Reply:
This bug is still there.